### ISSUE TITLE
Expand namespacing with additional features

### ISSUE THREAD

**Issue by b-studios** [effekt-lang/effekt#389]
PR #388 implements basic namespacing.  In the future it would be great to have support for additional features. In particular,

- [ ] control exporting selectively using visibility modifiers. Technically this could be implemented by annotating bindings with the modifier
    ```
    case class Bindings[F[_]](
      terms: Map[String, Set[F[TermSymbol]]], // terms can be overloaded
      types: Map[String, F[TypeSymbol]],
      captures: Map[String, F[Capture]],
      namespaces: Map[String, Bindings[F]]
    )
    
    enum Visibility[T] {
      case Public(binding: T)
      case Private(binding: T)
    
      def binding: T
    }
    ```
- [ ] separate include (that is,`include "foo/bar.effekt" as bar`) and import and 
- [ ] allow more fine-grained imports, i.e., `import foo::bar::{ x, y }`, renaming imports etc.
- [ ] add `export foo` for reexporting, again potentially with renaming
- [ ] add a full project mode where all (specified) files are always processed so includes are not necessary. Such a project mode could also configure other flags to the Effekt compiler, as well as declare dependencies etc.

Look at https://gist.github.com/jiribenes/8195243f56a454d3465b06f02e8bb5a2?permalink_comment_id=4872857#gistcomment-4872857 for discussions on designs.

**Comment by jiribenes** [comment]
Related to https://github.com/effekt-lang/effekt/issues/426

**Comment by jiribenes** [comment]
re: terminology, the last iteration of a proposal was https://gist.github.com/jiribenes/8195243f56a454d3465b06f02e8bb5a2?permalink_comment_id=4872857#gistcomment-4872857

**Comment by dvdvgt** [comment]
Yes, good point. I wasn't sure if everything here is up-to-date. Thanks for the link!

**Comment by jiribenes** [comment]
I think that, at least when the proposal came up, we were pretty confident in it, especially as it breaks down the whole namespacing/importing problem into four neat combinators with super clear rewriting semantics.
Then the actual import/module system is just syntactic sugar on top.
